### ISSUE TITLE
[FEATURE] Widening int mul (VPMULUDQ) and overflow checked int ops

### ISSUE THREAD

**Issue by chriselrod** [jfalcou/eve#1510]
You can see one approach implemented for 64 bit integers using highway here: https://gcc.godbolt.org/z/YWx3vaTET
This needs the `MulEven`/VPMULUDQ instruction.

The function multiplies two vectors elementwise, and returns `true` if there was overflow, `false` otherwise.
Ideally, we'd have optional overflow checks on `+` and `-` as well.

Here is a version without overflow checks: https://gcc.godbolt.org/z/x37f1hjxK

I am open to ideas for better implementations.
Overflow checked, AVX512 (unrolled 1x): https://bit.ly/3QbxbPV
No overflow check, AVX512 (unrolled 4x): https://bit.ly/3jOwGir
So it'd take around 22.4 (cascadelake) or 32 (tiger lake) clock cycles per 32 elements with overflow checks, vs only 6.5 or 12 cycles without.

A better AVX512 implementation probably would use `vpmullq` to compute the product, and use `vpmuludq` only for the overflow check. Especially on Zen4, which has faster `vpmullq` than the Intel CPUs.

Because AVX2 is forced to use `vpmuludq` anyway, the performance penalty is much smaller.
Overflow: https://bit.ly/3WYlcaG
No overflow: https://bit.ly/3Gbmfxc

The application I have in mind is something like https://dl.acm.org/doi/10.1145/3485539
E.g., using checked arithmetic when doing polyhedral calculations, and then in case of overflow either returning without applying any transformations to a loop (much better than applying invalid transforms!), or possible falling back to a wider integer type.

On that note, 128 bit integer support would also be nice to have.

**Comment by jfalcou** [comment]
Hi!
Thanks for looking at EVE.

We have had discussion around widening operations quite a few times and we are pretty open to support them.
Your base code looks OK from afar, we can have a look at implementing them.
The fun idea is to have the operation returns the overflow status, this makes for a very interesting API.

FOr 128 bits integer, this is a request we also have quite a bit but most SIMD ISA lacks the trans-element bitwise operation that may helps make proper operation. A SoA version (ie register of 128bits int stored as 4x 32bits registers) is maybe easier to support.

**Comment by chriselrod** [comment]
> Your base code looks OK from afar, we can have a look at implementing them.

I just realized I also need to check the high bits from `lohi` and `hilo` that I shift off, to make sure they're also all `0`.
I fixed it here: https://gcc.godbolt.org/z/vE41oYdoG
I'm also calculating the product itself for AVX512 through `vpmullq`.
Now that we can't avoid calculating `hilo` and `lohi`, it doesn't help Skylake-X/Cascadelake
`a*b = vpmullq`: https://bit.ly/3GfgYEE
`a*b = (lohi + hilo)<<32 + lolo`: https://bit.ly/3GCZROM
It also worsens performance on tiger lake, because it can use port 5 for integer additions, but not for multiplications.

It should help Zen4 though, given that their `vpmullq` is just as cheap as their `vpmuludq`.

To revert the change, just replace
```diff
- auto ab = hn::Lanes(d) == 8 ? a*b : BitCast(d, lolo + hi);
+ auto ab = BitCast(d, lolo + hi);
```

Thus, unfortunately, fixing the overflow check increased the cost to 26 (cascadelake) or 36-40 (tiger lake) cycles per 32 multiplications.

I also notice the compiler keeps the mask in a general purpose register, `kmov`ing the mask it generates from the comparison to a GRP (using the valuable port 0 to do so!), and `or`ing it there.
Seems like it'd have been reasonable to just keep the mask in mask register?

> The fun idea is to have the operation returns the overflow status, this makes for a very interesting API.

I'm not sure what exactly the internal API would have to look like. As a user, the perfect API would let me use arbitrary `eve::algo`s and get a flag back if so desired.
More realistically, probably something like `overflowing_mul`, `overflowing_abs` etc that can be used inside `eve::algo::[transform_]reduce`. One can make the overflow flag an element of the returned tuple (or return a tuple instead of a scalar).

> A SoA version (ie register of 128bits int stored as 4x 32bits registers) is maybe easier to support.

It's unfortunate that there aren't equivalent instructions for 64 bit ints.
For storing 128 bit integers, I think an SoA version storing 2x 64 bit ints may be fine. Because we can only multiply the even indices of a vector with vpmuludq, we still need bitshifts when using 32 bit ints.

**Comment by jfalcou** [comment]
@DenisYaroshevskiy can our current algo for transform and reduce handle a function that return a tuple ?

**Comment by DenisYaroshevskiy** [comment]
I think you'd have to write this wth for each since we can't 

```
using P = kumi::tuple<T, eve::logical<T>>;
eve::wide<P> // does not work
```

`eve::algo::for_each` is the best we can do here: https://godbolt.org/z/fvfcjMvo6

**Comment by jfalcou** [comment]
> ```
> using P = kumi::tuple<T, eve::logical<T>>;
> eve::wide<P> // does not work
> ```

Indeed, I think this may need to be fixed somehow as there is no real reason it can't.

**Comment by chriselrod** [comment]
I haven't tried it yet but watched your CppCon21 presentation -- doesn't it already work for custom types and complex numbers?
Tuple support would be nice, but would need users to pass in a custom `add` function, so it's not that far removed from defining one's own type.

**Comment by jfalcou** [comment]
Tuple support works, but there is a small oversight when using tuple of bool like value. This could be fixed somehow I guess.
I filled an issue so we can remember to solve it.

I think we can first give a go to having a widening mul support first then come back a bit later and add the value+check variant.

**Comment by DenisYaroshevskiy** [comment]
Btw: if you know what you want to write with intrinsics, it is very easy to integrate that with eve: all `eve::wide` and `eve::logical` are implicitly convertible to and from the underlying intrinsics.

**Comment by jfalcou** [comment]
#1512 has been merged so tuple with logical should be usable now.
